### PR TITLE
Reconstruct go ps: add kernelnew package, add IndexedTensor

### PR DESCRIPTION
**Pull request by lwpyr**
* add kernelnew package
* add IndexedTensor definition

### CODE REVIEW

**Comment by lwpyr** [comment]
The kernel code is just modified from the original one, so this is  not a large PR.

File: elasticdl/pkg/commonnew/tensor.go [hunk 1]
```diff
@@ -178,13 +178,35 @@ func (t *Tensor) SerializeToTensor() *tensor_go_proto.TensorProto {
 	}
 }
 
+// IndexedTensor : IndexedSlice in memory representation
+type IndexedTensor struct {
```
Comment (QiJune) [inline elasticdl/pkg/commonnew/tensor.go:182]: We could just follow proto message definition, name it to `IndexedSlices`

```go
type IndexedSlices struct {
    ConcatTensors *Tensor
    IDs []int64
}
```

File: elasticdl/pkg/kernelnew/capi/kernel_api.cc [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#include "kernel_api.h"
+
+#include <cmath>
+#include <eigen3/Eigen/Dense>
+
+void AddTo(float* a, float* b, long long size) {
```
Comment (QiJune) [inline elasticdl/pkg/kernelnew/capi/kernel_api.cc]: AddTo is not used, please remove it.

File: elasticdl/pkg/kernelnew/kernel.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	gradPtr := (*C.float)(unsafe.Pointer(&grad.Buffer[0]))
+	paramPtr := (*C.float)(unsafe.Pointer(&param.Buffer[0]))
+	length := len(grad.Buffer) / int(commonnew.DtypeSize[grad.Dtype])
+	C.SGD(gradPtr, paramPtr, C.float(lr), C.longlong(length))
+	return nil
+}
+
+// SparseSGD kernel
+func SparseSGD(grad *commonnew.IndexedTensor, param *commonnew.EmbeddingTable, lr float32) error {
```
Comment (QiJune) [inline elasticdl/pkg/kernelnew/kernel.go:22]: There are three cases:
- apply dense gradient to dense parameter
- apply sparse gradient to dense parameter. The gradient of native keras embedding layer is sparse, but the embedding parameter is dense.
- apply sparse gradient to sparse parameter 

So, we need an extra SGDKernel

File: elasticdl/pkg/kernelnew/kernel.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+// SparseAdam kernel
+func SparseAdam(grad *commonnew.IndexedTensor, param *commonnew.EmbeddingTable, m *commonnew.EmbeddingTable, v *commonnew.EmbeddingTable, lr float32, step int64, beta1 float32, beta2 float32, epsilon float32, amsgrad bool, maxSquare *commonnew.EmbeddingTable) error {
+	if grad.Dims[1] != param.Dim {
+		return fmt.Errorf("grad width is not equal to embedding dim")
+	}
+	for i, index := range grad.Indices {
+		if index == -1 {
+			continue
+		}
```
Comment (QiJune) [inline elasticdl/pkg/kernelnew/kernel.go]: Please remove the code

```go
if index == -1 {
    continue
}
```

File: elasticdl/pkg/commonnew/tensor.go (hunk 1 above)
Comment (lwpyr) [reply elasticdl/pkg/commonnew/tensor.go:182]:
```
Tensor -> TensorProto
IndexedSlice -> IndexedSliceProto
```
could this be better?

Comment (QiJune) [reply elasticdl/pkg/commonnew/tensor.go:182]: Since it's in memory representation, I think Proto is not needed.

Comment (QiJune) [reply elasticdl/pkg/commonnew/tensor.go:182]: And here, we could use `tensorflow.TensorProto` and `proto.IndexedSlices` directly. 

Pros:
- no need to define extra in memory data structures, `Tensor`/`IndexedSlices`
- no need to define extra serialize/deserialize functions, `SerializeTensor`/`DeserializeTensor`/`SerializeIndexedSlices`/`DeserializeIndexedSlices`

Cons:
- tensorflow.TensorShapeProto is not convenient to use directly, we need to add two extra function, `SetDim`/`GetDim`

File: elasticdl/pkg/kernelnew/kernel.go (hunk 4 above)
Comment (lwpyr) [reply elasticdl/pkg/kernelnew/kernel.go]: fixed

File: elasticdl/pkg/kernelnew/kernel.go (hunk 3 above)
Comment (lwpyr) [reply elasticdl/pkg/kernelnew/kernel.go:22]: fixed

File: elasticdl/pkg/kernelnew/capi/kernel_api.cc (hunk 2 above)
Comment (lwpyr) [reply elasticdl/pkg/kernelnew/capi/kernel_api.cc]: fixed

File: elasticdl/pkg/commonnew/tensor.go (hunk 1 above)
Comment (QiJune) [reply elasticdl/pkg/commonnew/tensor.go:182]: I find that it's not convenient to write codes without defining in memory data structures. So let's move on.

And for the struct name, How about

```go
type struct IndexedSlices {
    ConcatTensors *Tensor
    IDs []int64
} 
```

**Review by QiJune — APPROVED** [review]
Excellent!
